File: algoritmos/III_5_EnhancedParallelBlock.py
from concurrent.futures import ThreadPoolExecutor
import threading

def block_multiply_section(matrizA, matrizB, matrizRes, start, end, size2, lock):
    size1 = len(matrizA)
    for i1 in range(start, end, size2):
        for j1 in range(0, size1, size2):
            for k1 in range(0, size1, size2):
                for i in range(i1, min(i1 + size2, size1)):
                    for j in range(j1, min(j1 + size2, size1)):
                        for k in range(k1, min(k1 + size2, size1)):
                            # Usar el lock para garantizar que solo un hilo pueda actualizar la matrizRes a la vez
                            with lock:
                                matrizRes[i][j] += matrizA[i][k] * matrizB[k][j]

def alg_III_5_Enhanced_Parallel_Block(matrizA, matrizB, size1, size2):
    matrizRes = [[0.0 for _ in range(size1)] for _ in range(size1)]
    mid_point = size1 // 2
    lock = threading.Lock()  # Crear un Lock para sincronización
    with ThreadPoolExecutor(max_workers=2) as executor:
        # Dividir el trabajo en dos tareas paralelas
        executor.submit(block_multiply_section, matrizA, matrizB, matrizRes, 0, mid_point, size2, lock)
        executor.submit(block_multiply_section, matrizA, matrizB, matrizRes, mid_point, size1, size2, lock)
    return matrizRes

def multiply(matrizA, matrizB):
    N = len(matrizA)
    P = len(matrizB[0])
    matrizRes = alg_III_5_Enhanced_Parallel_Block(matrizA, matrizB, N, P)
    return matrizRes

File: AlgoritmosKotlin/src/main/kotlin/algoritmos/III_5_EnhancedParallelBlock.kt
package algoritmos

import java.util.concurrent.Executors
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class III_5_EnhancedParallelBlock {
    fun blockMultiplySection(
        matrizA: Array<DoubleArray>,
        matrizB: Array<DoubleArray>,
        matrizRes: Array<DoubleArray>,
        start: Int,
        end: Int,
        size2: Int,
        lock: ReentrantLock
    ) {
        val size1 = matrizA.size
        for (i1 in start until end step size2) {
            for (j1 in 0 until size1 step size2) {
                for (k1 in 0 until size1 step size2) {
                    for (i in i1 until minOf(i1 + size2, size1)) {
                        for (j in j1 until minOf(j1 + size2, size1)) {
                            for (k in k1 until minOf(k1 + size2, size1)) {
                                lock.withLock {
                                    matrizRes[i][j] += matrizA[i][k] * matrizB[k][j]
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    fun algIII5EnhancedParallelBlock(
        matrizA: Array<DoubleArray>,
        matrizB: Array<DoubleArray>,
        size1: Int,
        size2: Int
    ): Array<DoubleArray> {
        val matrizRes = Array(size1) { DoubleArray(size1) }
        val midPoint = size1 / 2
        val lock = ReentrantLock() // Crear un Lock para sincronización
        val executor = Executors.newFixedThreadPool(2)

        try {
            // Dividir el trabajo en dos tareas paralelas
            val task1 = Runnable {
                blockMultiplySection(matrizA, matrizB, matrizRes, 0, midPoint, size2, lock)
            }
            val task2 = Runnable {
                blockMultiplySection(matrizA, matrizB, matrizRes, midPoint, size1, size2, lock)
            }

            executor.submit(task1)
            executor.submit(task2)
        } finally {
            executor.shutdown() // Asegurarse de cerrar el executor después de completar las tareas
            while (!executor.isTerminated) {
                // Esperar hasta que todas las tareas hayan terminado
            }
        }

        return matrizRes
    }

    fun multiply(matrizA: Array<DoubleArray>, matrizB: Array<DoubleArray>): Array<DoubleArray> {
        val N = matrizA.size
        val P = matrizB[0].size
        return algIII5EnhancedParallelBlock(matrizA, matrizB, N, P)
    }
}
